File: examples/vector_quick_start.rs
use aletheiadb::index::vector::temporal::TemporalVectorConfig;
use aletheiadb::{AletheiaDB, DistanceMetric, HnswConfig, SimilarityQuery, properties};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let db = AletheiaDB::new().unwrap();

    // Enable vector indexing with temporal support
    db.vector_index("embedding")
        .hnsw(HnswConfig::new(384, DistanceMetric::Cosine))
        .temporal(TemporalVectorConfig::default())
        .enable()?;

    let embedding = vec![0.1f32; 384];

    // Store node with embedding - automatically indexed!
    // Note: We convert the vector to a slice for the properties! macro
    let doc_id = db.create_node(
        "Document",
        properties! {
            "title" => "Introduction to Rust",
            "embedding" => &embedding[..],
        },
    )?;

    // Create another similar node so we have something to find
    let _doc2_id = db.create_node(
        "Document",
        properties! {
            "title" => "Rust for Beginners",
            "embedding" => &embedding[..], // Same embedding for max similarity
        },
    )?;

    // Find similar nodes
    // Note: the similarity search excludes the query node itself from results
    let similar = db.similarity_search(SimilarityQuery::from_node(doc_id).k(10))?;

    println!("Similar nodes: {:?}", similar);

    Ok(())
}

File: python/examples/vector_quick_start.py
"""Vector search quickstart.

Run with: python examples/vector_quick_start.py
"""
from aletheiadb import (
    AletheiaDB,
    DistanceMetric,
    HnswConfig,
    enable_vector_index,
    find_similar,
)


def main() -> None:
    db = AletheiaDB()
    enable_vector_index(db, "embedding", HnswConfig(4, DistanceMetric.COSINE))

    rust = db.create_node("Document", {
        "title": "Rust by Example",
        "embedding": [1.0, 0.0, 0.0, 0.0],
    })
    python = db.create_node("Document", {
        "title": "Python Programming",
        "embedding": [0.9, 0.1, 0.0, 0.0],
    })
    cooking = db.create_node("Document", {
        "title": "Cooking 101",
        "embedding": [0.0, 0.0, 1.0, 0.0],
    })

    print("Documents most similar to 'Rust by Example':")
    for node_id, score in find_similar(db, rust, k=3):
        title = db.get_node(node_id).properties["title"]
        print(f"  {title!r}: score={score:.4f}")


if __name__ == "__main__":
    main()
